### PR TITLE
when it's test failure reproduction run read a specified json file and rerun a list of module for specified browser.

### PR DESCRIPTION
**Pull request by choheekim**

### CODE REVIEW

File: lib/commands/exam.js [hunk 1]
```diff
@@ -332,6 +334,44 @@ module.exports = TestCommand.extend({
     }
   },
 
+  /**
+   * Adds an object which maps a list of module from browser id to config
```
Comment (dcombslinkedin) [inline lib/commands/exam.js]: list of modules, not list of module.

File: lib/commands/exam.js [hunk 2]
```diff
@@ -332,6 +334,44 @@ module.exports = TestCommand.extend({
...
+    const replayBrowser = commandOptions.replayBrowser;
+
+    let testMappingObj = {}
+
+    try {
+      // Read the replay execution json file.
+      const testMappingRawData = fs.readFileSync(filePath);
+      testMappingObj = JSON.parse(testMappingRawData);
+    } catch (err) { 
+      throw new Error(err);
```
Comment (dcombslinkedin) [inline lib/commands/exam.js]: I'm not sure I'd directly throw this error. I might wrap it in an error that says something like "error reading replay execution JSON file", then the error string from this error. It's clearer to the user that way.

File: lib/commands/exam.js [hunk 3]
```diff
@@ -332,6 +334,44 @@ module.exports = TestCommand.extend({
...
+    }
+
+    const browserModuleMapping = ((replayBrowser, testMappingObj) => {
+      const browserModuleMappingObj = {};
+      for (const idx in replayBrowser) {
+        const browserId = replayBrowser[idx];
+        const listOfModule = testMappingObj[browserId];
```
Comment (dcombslinkedin) [inline lib/commands/exam.js]: listOfModules, not listOfModule

File: lib/commands/exam.js [hunk 4]
```diff
@@ -332,6 +334,44 @@ module.exports = TestCommand.extend({
...
+
+        if(!listOfModule) {
+          throw Error ('browser id doesn\t exist in the file.');
```
Comment (dcombslinkedin) [inline lib/commands/exam.js]: ID, not id. Also, the error message needs to be more clear about what it means, like 'the file' should really say which file it's talking about, and 'browser id', should be replaced with something saying what the ID is you're looking for (presumably it was specified by the caller?)

File: lib/commands/exam.js [hunk 5]
```diff
@@ -343,15 +383,27 @@ module.exports = TestCommand.extend({
 
     events['set-modules-queue'] = function (modules) {
       //When `load-balance` option is valid we want to have one static list of modules on server side to send a module path to browsers.
-      const proto = Object.getPrototypeOf(this);
-      if (!proto.moduleQueue) {
+      const proto = Object.getPrototypeOf(this)
+      const browserId = (() => {
```
Comment (dcombslinkedin) [inline lib/commands/exam.js:395]: Same comment as in the other file - you don't really need a function for this, do you?

File: lib/commands/exam.js [hunk 6]
```diff
@@ -364,15 +416,24 @@ module.exports = TestCommand.extend({
...
+      // if this run is for failure reproduction it's not need to create moduleMap.
+      if (!replayExecution) {
+        proto.moduleMap[browserId] = proto.moduleMap[browserId] || [];
+        proto.moduleMap[browserId].push(moduleName);
+      }
     }
 
     events['after-tests-complete'] = function () {
       const proto = Object.getPrototypeOf(this);
       const fileName = 'test-execution-%TIMESTAMP%.json'.replace('%TIMESTAMP%', Date.now());
```
Comment (dcombslinkedin) [inline lib/commands/exam.js]: Same comment as other file:    fileName = `test-execution-${Date.now()}.json`;

**Review by dcombslinkedin — CHANGES_REQUESTED** [review]
minor stylistic issues, but otherwise okay.

**Comment by scalvert** [comment]
Can we open these against `ember-cli/ember-exam` instead of this repo. It's going to be more work/time to PR against this repo, then have to PR again against upstream. Additionally, we want to get eyes on this code as early as possible. 

I'd suggest we be very explicit, in the PR description for _all_ the PRs that are related, what the sequence is for landing these changes.

For example - possible PR titles and description:

Title: [Part 1] - Load balancing Ember Exam - add weighted option
Description: Adds weighted option to ember exam to provide better utilization of browser parallelization.

This change is Part 1 in a 5 part series:
- Part 1 - add weighted option
- Part 2 - do something else
- Part 3 - do the part 3 thing
- etc.

**Pull request closed without merging**